### ISSUE TITLE
MongoDbRepository fails handling of large datasets

### ISSUE THREAD

**Issue by esmn** [SharpRepository/SharpRepository#211]
On the core 2.0 project, we are using MongoDbRepository. Some of our queries return large size of data (over 100 MB). We realized that we are getting an error which says: "Command aggregate failed: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting. Aborting operation. Pass allowDiskUse:true to opt in..".

After quick search I found that MongoDb restricts return data size to 100 MB. MongoDb documetnts says: "Pipeline stages have a limit of 100 megabytes of RAM. If a stage exceeds this limit, MongoDB will produce an error. To allow for the handling of large datasets, use the allowDiskUse option to enable aggregation pipeline stages to write data to temporary files.".

To enable returning large size of data, we need to pass "allowDiskUse: true" option in our query. Is there any way to pass that option to MongoDb with SharpRepository? 

https://docs.mongodb.com/manual/core/aggregation-pipeline-limits/#agg-memory-restrictions

**Comment by omarpiani** [comment]
first question is.. how can be done using C# driver?

**Comment by omarpiani** [comment]
How you make aggregation in SharpRepository?  For me actual aggregations in SharpRepository are made in list afrer getting all documents from mongo.

For example Sum is FindAll(criteria, selector).ToList().Sum()...

It's really strange you get aggregation errors

**Comment by omarpiani** [comment]
Probably driver generates aggregates automatically while mapping to your model.. can you post your model?

**Comment by esmn** [comment]
Thank you for your quick response.

We would like to order it first then get dataset as a list, so our code is like this:
`_repository.AsQueryable().Where(predicate).OrderByDescending(keySelector).ToList()`

Based on your suggestion, if we get dataset first then order it, it works fine.
`_repository.AsQueryable().Where(predicate).ToList().OrderByDescending(keySelector)`

But we would like to aggregate the dataset in MongoDb side, not in our side. The reason we want it to be handle from MongoDb is that, the server which has MongoDb is more powerfull than the API server. Is there any way to let MongoDb to handle it?

**Comment by omarpiani** [comment]
can you log the query sent to the  server?

**Comment by esmn** [comment]
`aggregate([{ "$match" : { "PublishTime" : { "$gt" : ISODate("2018-04-30T21:00:00Z") } } }, { "$match" : { "PublishTime" : { "$lt" : ISODate("2018-05-30T21:00:00Z") } } }, { "$sort" : { "PublishTime" : -1 } }])`

**Comment by omarpiani** [comment]
mmhhhh I'm not expert on mongo.. seems filtering between 2 dates and sorting... how/what is aggregating?

**Comment by esmn** [comment]
I'm also not very experienced on mongo. When I search aggregation pipeline stages on mongo, both of match and sort are stages in aggregation pipeline. I tried some other stages in aggregation pipeline, such as count, limit, skip, group and none of them throw an exception. In my case, the only stage which throws an exceeded memory limit exception is shorting. I believe that when mongo is shorting my dataset, it will need a memory over 100 MB, that's why it throws an exception. Other stages return same dataset but they don't throw an exception. Again in my case, it's not about my returning dataset size, it's about when mongo shorting my dataset, it will need over 100 MB RAM. But I'm not sure how we can let SharpRepository to pass "allowDiskUse: true" option to MongoDb.

**Comment by omarpiani** [comment]
Whoa.. with some luck I found that mongo driver support this:
`collection.AsQueryable(new AggregateOptions { AllowDiskUse = true } )`

Can you make a fork and verify that it works?
https://github.com/SharpRepository/SharpRepository/blob/7acbc36014efc9213a37b1f79037bdbf82dabdc1/SharpRepository.MongoDbRepository/MongoDbRepositoryBase.cs#L119
 if yes I can put a FetchStrategy parameter with ability to activate this flag

**Comment by esmn** [comment]
Awesome!! It just worked perfect. Looking forward you to update it. Thank you for your help.

**Comment by omarpiani** [comment]
Released MongoDbRepository 2.0.6-beta1 with this update

You have to use with
```
 var fetchStrategy = new MongoDbFetchStrategy<Order>
 {
     AllowDiskUse = true
 };
 var all = repo.GetAll(fetchStrategy);
```

**Comment by esmn** [comment]
Well, It works fine if GetAll, MongoDbFetchStrategy and SortingOptions used together. But I would like to use it like:

` var all = repo.AsQueryable(fetchStrategy);`

Can you let the AsQueryable get FetchStrategy parameter with ability to activate AllowDiskUse flag?

**Comment by omarpiani** [comment]
Mh... it's a needed improvement for Mongo.. but this touches the SharpRepository.Repository and propagates this to Ef/EfCore repositories.. 

For now I put to MongoDbRepository but can't be visible if you use IRepository and you need to cast it to MongoDbRepository.

**Comment by omarpiani** [comment]
Released 2.0.6.1-beta2

**Comment by esmn** [comment]
Awesome! Thank you for your help and quick responses.

**Comment by omarpiani** [comment]
is ready for stable release @esmn ?

**Comment by esmn** [comment]
Didn't face any problem so far. I believe, it's ready for stable one.

**Comment by omarpiani** [comment]
Released MongoDbRepository 2.0.6.2
